### PR TITLE
Allow style sharing for elements with ids as long as the ID is not being used for styling

### PR DESCRIPTION
**Pull request by bzbarsky**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix part of the problem we're seeing in https://bugzilla.mozilla.org/show_bug.cgi?id=1367862

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because the only impact is on performance and memory.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17055)

### CODE REVIEW

File: components/style/sharing/checks.rs [hunk 1]
```diff
@@ -154,3 +155,28 @@ pub fn revalidate<E>(element: E,
 
     for_element == for_candidate
 }
+
+/// Checks whether we have rules for either of the two ids.
+#[inline]
+pub fn have_rules_for_ids(shared_context: &SharedStyleContext,
+                          element_id: Option<Atom>,
```
Comment (emilio) [inline components/style/sharing/checks.rs]: This could better take a reference for its arguments, as discussed on IRC.

**Review by emilio — APPROVED** [review]
r=me with that bit. The patch you had on IRC seemed fine.

Thanks!

**Comment by bzbarsky** [comment]
@bors-servo retry

Looks like at least https://github.com/servo/servo/issues/16627 intermittently failing.

**Comment by jdm** [comment]
The filtered log from the filter-intermittents log shows a bunch of tests that used to fail are now reliably passing.

**Comment by bzbarsky** [comment]
Right.  That's actually due to a bug this patch introduces.  Specifically, with this patch, this testcase:

    <!DOCTYPE html>
    <style>
      form { color: red; }
      #foo form { background-color: green }
    </style>
    <div>
      <form></form>
    </div>
    <div id="foo">
      <form>This should be green</form>
    </div>

fails.

I'm sorting through whether I can construct a similar testcase that fails without this patch.

**Comment by bzbarsky** [comment]
For those following along, some notes:

First, the testcase fails because the patch violates an invariant that the style sharing cache depends on.  That invariant goes something like this: "Every simple selector must either trigger revalidation or have its matching result guaranteed by the pre-checks done in test_candidate".  With this patch, the pre-checks no longer guarantee matching results for id selectors, because they only check that the id is not used in the _rightmost_ compound selector of a complex selector.  So in the testcase above, the two divs share style (because the pre-checks don't prevent them from doing that), and then the two forms end up sharing style (because all pre-checks pass for them, their parents are sharing style, and no revalidation gets done).

Conclusion: if we want to do this optimization in this form, we need to make selectors involving an id revalidation selectors.  But we only need to do that if they're not in the rightmost compound selector, really.  Not sure how best to do that in our current revalidation setup.

Second, if you reverse the order of those two divs, and add text to the form without an id, you will notice that sharing is no longer happening.  That's because relations_are_shareable does not insert an element into the style cache if it was matched by a selector that had an id simple selector somewhere in it.  This seems pretty weird to me; in particular it means that a `#foo *` selector on a page with `<html id="foo">` will completely turn off the cache.  I _think_ that if we make ids revalidation selectors we can remove this bit.

Still digging into some other weirdness I'm seeing.

**Comment by bzbarsky** [comment]
Oh, and the patch has another bug: it checks element_id both times in have_rules_for_ids.  The second check should be for candidate_id.   I'll fix that.

**Comment by bholley** [comment]
> But we only need to do that if they're not in the rightmost compound selector, really.

Well, this _and_ if the id never appears in the rulehash, since that's the only situation where we do this optimization. This additional constraint should cut down the number of revalidation selectors significantly.

> Not sure how best to do that in our current revalidation setup.

IMO, we should collect selectors with a non-rightmost ID selector into a SmallVec in note_selector. Afterwards, we should iterate those and check the corresponding id against the primary SelectorMap, and insert the additional revalidation selectors as appropriate.

Note that I'm going to parallelize stylist::update in bug 1362538. But this shouldn't be too much of a problem, it would just be a final sequential step after all the SelectorMaps have been built.

**Comment by bholley** [comment]
>  I think that if we make ids revalidation selectors we can remove this bit.

Yes, this is one of those dumb carryover behaviors from the old cache model. We should fix it.

**Comment by bzbarsky** [comment]
So I implemented all of the above, and that still doesn't fix the testcase in https://github.com/servo/servo/pull/17055#issuecomment-304426771 with stylo and STYLO_THREADS=1 (so we actually trigger the sharing).

That's because there is one more not quite explicitly stated invariant around the ComputedValues cache.  Because revalidate() uses the passed-in Bloom filter for revalidation selector matching for both "element" and "candidate", we depend on that filter giving identical answers for all revalidation selectors for both "element" and "candidate".

In the testcase above, that invariant is not maintained with my patches.  The Bloom filter is in the state for ... I'm not actually sure what.  The upshot is that the "#foo form" revalidation selector does not match _either_ form when we're checking whether we can do sharing, and that's because the Bloom filter rejects it.  I would have expected the filter to be in the state for the element, that is the second form in the document, but we seem to be thinking that the second form does _not_ match `"#foo form"`, because of the Bloom filter.  I haven't checked the details of why yet.  Need to sort this out before I can come up with a solution; it would have made a lot more sense if the Bloom filter results for the cached thing were wrong than if they're wrong for the element we're matching.

**Comment by emilio** [comment]
On May 28, 2017 4:01 AM, Boris Zbarsky <notifications@github.com> wrote:So I implemented all of the above, and that still doesn't fix the testcase in #17055 (comment) with stylo and STYLO_THREADS=1 (so we actually trigger the sharing).
That's because there is one more not quite explicitly stated invariant around the ComputedValues cache.  Because revalidate() uses the passed-in Bloom filter for revalidation selector matching for both "element" and "candidate", we depend on that filter giving identical answers for all revalidation selectors for both "element" and "candidate".Ouch, I has forgotten about the fact that we passed down the bloom filter down the revalidation code... That, plus cousin sharing, surely makes this somewhat annoying.
In the testcase above, that invariant is not maintained with my patches.  The Bloom filter is in the state for ... I'm not actually sure what.  The upshot is that the "#foo form" revalidation selector does not match either form when we're checking whether we can do sharing, and that's because the Bloom filter rejects it.  I would have expected the filter to be in the state for the element, that is the second form in the document, but we seem to be thinking that the second form does not match "#foo form", because of the Bloom filter.  I haven't checked the details of why yet.  Need to sort this out before I can come up with a solution; it would have made a lot more sense if the Bloom filter results for the cached thing were wrong than if they're wrong for the element we're matching.That's weird, though it's because in compute_style (traversal.rs) we only deal with the bloom filter on the MatchAndCascade branch.I bet this is somewhat wallpapered because a combination of the bail out/clear the cache on parent mismatch bits.You are receiving this because you were assigned.Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by emilio** [comment]
Err, I had to edit the above comment to fix some pretty messed up quoting. Anyway, the explanation of why the bloom filter is not setup when we do the sharing stuff is because when originally written there was no bloom filter in the style sharing code at all (doing some digging seems it was added in 1663bf6e5dbf73ffdca727371193231f1a25d1a3, r=emilio, of course :/).

**Comment by bholley** [comment]
So, this is definitely annoying, however there are some mitigations:
* This doesn't affect situations where we already matched revalidation selectors for the cache entry back when we were computing its style.
* This only affects cousin sharing, not sibling sharing.

So I think we should just check the validity of the bloom filter when lazily computing revalidation selectors for cache entries, and if it's not valid, don't pass it. Should be a one-line change.

**Comment by bzbarsky** [comment]
OK.  I'm going to punt the smart invalidation selector bits to https://bugzilla.mozilla.org/show_bug.cgi?id=1369611 in the interests of actually getting this landed.  I need to read a bit more about how we actually represent Atom and selectors whatnot to have a good idea of how best to avoid various refcounting and copying.

**Comment by bzbarsky** [comment]
@emilio the third commit is pretty much what you reviewed before; the other three commits fix up the various other things mentioned above.  I think this should be good to go; mind reviewing?

**Comment by bzbarsky** [comment]
OK, those patches still fail http://searchfox.org/mozilla-central/rev/972b7a5149bbb2eaab48aafa87678c33d5f2f045/layout/reftests/w3c-css/submitted/variables/variable-reference-12.html

The reason they fail that test is that the style for the first `<p>` goes into the style sharing cache, as expected.  But the style for the second `<p>` ends up sharing it, because we don't add the `#a:before` to revalidation selectors, because my check for whether an id is in some rulehash only checks the "no pseudos" rulehash, so if we added that to revalidation selectors we'd end up with length mismatches when pulling things out of the revalidation selector rulehash.

I guess we could make the check for id being in some rulehash check all pseudo rulehashes in addition to the "no pseudos" one... I'll think a bit about whether we can do better.

File: components/style/sharing/mod.rs [hunk 2]
```diff
@@ -92,19 +92,40 @@ impl ValidationData {
...
     {
         if self.revalidation_match_results.is_none() {
+            // The bloom filter may already be set up for our element.
+            // If it is, use it.  If not, we must be in a candidate
+            // (i.e. something in the cache), and the element is one
+            // of our cousins, not a sibling.  In that case, we'll
+            // just do revalidation selector matching without a bloom
+            // filter, to avoid thrashing the filter.
+            let bloom_to_use = if bloom_known_valid {
+                debug_assert_eq!(bloom.current_parent(),
```
Comment (emilio) [inline components/style/sharing/mod.rs]: You can use `bloom.assert_complete(element)` if you want a bit more precise checking... Though probably this is enough.

File: components/style/sharing/mod.rs [hunk 3]
```diff
@@ -228,6 +251,7 @@ impl<E: TElement> StyleSharingTarget<E> {
             self.element,
             stylist,
             bloom,
+            true,
```
Comment (emilio) [inline components/style/sharing/mod.rs]: Please use `/* bloom_known_valid = */ true`

File: components/style/sharing/mod.rs [hunk 4]
```diff
@@ -146,16 +167,18 @@ impl<E: TElement> StyleSharingCandidate<E> {
...
         &mut self,
         stylist: &Stylist,
-        bloom: &BloomFilter,
+        bloom: &StyleBloom<E>,
     ) -> &BitVec {
         self.validation_data.revalidation_match_results(
             *self.element,
             stylist,
             bloom,
+            false,
```
Comment (emilio) [inline components/style/sharing/mod.rs]: ditto re. the comment on the bool arg. Alternatively, you can use a binary enum like:

```rust
enum BloomFilterValidity {
    KnownValid,
    MaybeInvalid,
}
```

Or something like that.

**Review by emilio — COMMENTED** [review]
Just a partial review looking through the code. Looks good, though last comment needs of course to be fixed.

We can keep a `FnvHashSet<Atom>` with the id's that appear on selectors, that we could populate while we scan them, or something like that, I guess... Haven't thought through it a lot though.

**Comment by bholley** [comment]
We can use a bloom filter instead of an FnvHashSet.

**Comment by bzbarsky** [comment]
@emilio review ping?

That said, I'm not sure how much the visitor bits here will bitrot as a result of https://bugzilla.mozilla.org/show_bug.cgi?id=1370107 and whether I should wait for that to land first...

**Comment by bholley** [comment]
Per IRC discussion, that work will not bitrot these patches.

File: components/style/stylist.rs [hunk 5]
```diff
@@ -1064,6 +1078,14 @@ impl Stylist {
         debug!("push_applicable_declarations: shareable: {:?}", context.relations);
     }
 
+    /// Given an id, returns whether there might be any rules for that id in any
+    /// of our rule maps.
+    #[inline]
+    pub fn may_have_rules_for_id(&self,
+                                 id: &Atom) -> bool {
```
Comment (emilio) [inline components/style/stylist.rs]: nit: This fits in 80 chars I think.

File: components/style/sharing/mod.rs [hunk 6]
```diff
@@ -4,6 +4,67 @@
...
+//! selectors any time there are two elements that could pass all the up-front
+//! checks but match differently against some ComplexSelector in the selector.
+//! If that happens, then they can have descendants that might themselves pass
+//! the up-front checks but would have different matching results for the
+//! selector in question.  In this case, "descendants" includes pseudo-elements,
+//! so there is a single selector map of revalidation selectors that includes
+//! both selectors targeting element and selectors targeting pseudo-elements.
+//! This relies on matching an element against a pseudo-element-targeting
+//! selector being a sensible operation that will effectively check whether that
+//! element is a matching originating element for the selector.
```
Comment (emilio) [inline components/style/sharing/mod.rs]: Thanks for writing this :)

**Review by emilio — APPROVED** [review]
r=me

**Pull request merged**